draft-nro-bulk-rdap.md

```diff
@@ -48,184 +48,294 @@ feature of this service.
-JSON data for a single RDAP object class.
+The data returned for a Bulk RDAP request ((#bulk_rdap_url)) is a sequence of JSON objects, each followed by a newline
+('\n') character. Such a sequence is commonly known as [JSON Lines](https://jsonlines.org/), or newline-delimited JSON,
+and makes bulk data streaming computationally efficient for the clients.
+The first object in the returned sequence provides the metadata information for the bulk data, and the following objects
+are data objects for one or more RDAP object classes ([@!RFC9083, section 5]).
 
 ## Metadata
 
-The following JSON members for metadata MUST be included:
-
-* "rdapConformance" -- An array of strings ([@!RFC9083, section 4.1]) to signal the RDAP extensions the data is based
-  on.
-* "serial" -- An unsigned 32-bit number for sequencing the data snapshots. It is RECOMMENDED to follow the serial
-  number-wrapping arithmetic from [@!RFC1982].
-
-## Data for an Object Class
-
-The JSON data for an RDAP object class is an "objects" member -- a JSON array comprising all the objects for that class.
-It is RECOMMENDED that an RIR provide bulk data for IP Network, Autonomous System Number, Domain, Nameserver, and Entity
-object classes ([@!RFC9083, section 5]).
-
-Furthermore, within the JSON array:
-
-* The "self" links for each primary object and the secondary objects it contains MUST be included to have parity with
-  the "ref" element in the Bulk Whois.
-* For compactness, it is NOT RECOMMENDED to include details for a secondary object beside its "self" link, "handle" when
-  defined, and relationship to the primary object (e.g., using the "roles" member in an Entity object to relate to an IP
-  Network object).
-* An "rdapConformance" member MAY be included at the top of a primary object if the RDAP extensions used to produce it
-  are different from those listed in the "rdapConformance" member for metadata.
-
-## Example
-
-Here is an elided example of a JSON object containing bulk data for the IP Network object class:
+The metadata JSON object MUST include the following members:
+
+* "extensionId" -- the "nroBulkRdap1" string (see (#rdap_extensions_registry))
+* "versionId" -- a version 4 Universally Unique IDentifier (UUID, [@!RFC9562, section 5.4]) string that MUST be the same
+  for all the bulk data generated for various RDAP object classes at a particular interval (say, on a particular day)
+* "producer" -- a string identifying the registry that produced the bulk data, with possible values of "AFRINIC",
+  "APNIC", "ARIN", "LACNIC", "RIPE NCC", or an agreed-upon string literal for a registry at the national or local level
+* "productionDate" -- a string containing the date and time with the UTC offset ([@!RFC3339]) of the producer,
+  indicating when the bulk data being returned was produced
+* "objectCount" -- a positive number (greater than 0) representing the count of data objects following the metadata
```

Comment:
Why does this have to be greater than zero?

Reply:
By saying "greater than zero", I was trying to ensure that there is at the least one data object. That said, could there be a use case for sending a metadata object without any data objects?

Reply:
I think there is a use case for this for object-class-specific files where the registry doesn't host any data, e.g. `nameserver` for RIRs, and `ip`/`autnum` for name service operators.  Being able to see the assertion that there are no objects in this class may be helpful for consumers.

Reply:
I see your point about asserting 0 data objects for these scenarios. But, won't a 501 Not Implemented suffice here? Also, if we send back 0 `objectCount`, what do we say in `productionDate`?

Reply:
There could be situations where an object is supported but none exists.  Admittedly unlikely, but in Whois/IRR e.g., we have several supported object types with very few objects (e.g. there are only four rtr-set objects and five filter-set objects).  As far as `productionDate` goes, I think the text for that should be something like "the time at which the dump/snapshot was taken from the database", so as to cover cases like this one.

Reply:
Yah, I thought about that but you are right there could be a non-zero probability for this scenario.